### PR TITLE
fix CBN in error state undo/redo crash.

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

I found that undoing a modification to a code block already in an error state to another error state would add more outputPorts to that node - these ports are not visible but they do impact the saveAndDeleteConnector function which would then throw when trying to add ports with the same name to the dictionary of connections to save.

So if the code block is in an error state, we clear all the ports before deserializing any more from the runtime saved last good state.

I have added 2 tests and run the self serve CI.

Please note that this PR should be reviewed after 
https://github.com/DynamoDS/Dynamo/pull/8590

as those changes have been merged here.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang 
@ColinDayOrg

### CODE REVIEW

File: src/DynamoCore/Graph/Nodes/CodeBlockNode.cs [hunk 1]
```diff
@@ -390,15 +390,19 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
...
                 inputPortHelpers.Select(x => x.ReadString("name", String.Empty))
                     .Where(y => !string.IsNullOrEmpty(y))
                     .ToList();
             SetInputPorts();
 
-           
+            // if we're in an error state - clear the output ports before we try adding more.
+            if (IsInErrorState)
+            {
+                OutPorts.RemoveAll((p) => { return true; });
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Nodes/CodeBlockNode.cs]: this is the only real change.

File: test/DynamoCoreTests/CodeBlockNodeTests.cs [hunk 2]
```diff
@@ -310,6 +310,65 @@ public void CodeBlockConnectionsRemainAfterUndo()
             Assert.AreEqual(3, codeBlockNode2.CachedValue.Data);
 
         }
+        [Test]
+        public void UndoRedoCodeBlockErrorStateDoesNotCrashOutputs()
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/CodeBlockNodeTests.cs]: I verified this test was broken before the fix.

File: test/DynamoCoreTests/CodeBlockNodeTests.cs [hunk 3]
```diff
@@ -1280,6 +1339,7 @@ public void TestImperativePropertyAccessing()
         }
 
         [Test, Category("Failure")]
+        [Ignore("Test Loops Forever. Danger.")]
```
Comment (mjkkirschner) [inline test/DynamoCoreTests/CodeBlockNodeTests.cs]: @aparajit-pratap I added ignore to this test so that running dynamoCore tests locally is possible.

**Comment by ColinDayOrg** [comment]
LGTM

**Pull request merged**